### ISSUE TITLE
Soft mandatory labeling for release notes 

### ISSUE THREAD

**Issue by malfet** [pytorch/pytorch#68459]
Composing release notes has historically been a very time consuming process. Until now, we would have one dedicated engineer from the PyTorch Core team manually categorize all the commits (usually about 3-3.5k commits)  every release, which is usually about a week's worth of full-time work. 

In PyTorch 1.10, we made a significant improvement, by auto-categorizing about 30% (1k) of the 3.5k commits using [this](https://github.com/pytorch/pytorch/pull/64929) script that labeled PRs based on the files changed. 

Now, we are proposing to move to an upfront labeling system, where all PR authors categorize and topic each PR upfront, which will make compiling release notes much easier at the time of release with very little upfront work. This would mean that in the future releases, categorizing of PRs for compiling release notes will be solely done based on the category labels. 

Inspired by https://github.com/pytorch/vision/pull/4618 we propose to add GitHub workflow/Probot path that would remind authors to assign *one* of the following category labels to PRs:

- `release notes: mobile`
- `release notes: jit`
- `release notes: visualization`
- `release notes: onnx`
- `release notes: quantization`
- `release notes: rocm`
- `release notes: cuda`
- `release notes: benchmark`
- `release notes: profiler`
- `release notes: performance_as_product`
- `release notes: package/deploy`
- `release notes: composability`
- `release notes: releng`
- `release notes: fx`
- `release notes: vulkan`
- `release notes: nn`
- `release notes: cpp`
- `release notes notes: python_frontend`
- `release notes: complex`
- `release notes: autograd`
- `release notes: build`
- `release notes: memory format`
- `release notes: foreach_frontend`
- `release notes: dataloader`
- `release notes: linalg_frontend`
- `release notes: Meta API`
- `release notes: sparse`
- `release notes: distributed (c10d)`
- `release notes: distributed (rpc)`
- `release notes: distributed (ddp)`
- `release notes: distributed (pipeline)`
- `release notes: distributed (fsdp)`
- `release notes: distributed (sharded)`

Please leave a comment, with additional category labels if you are working on a new feature or want to refine an existing category label into more specific category labels to distribute the work or for other reasons. Example., at the time of writing this, `distributed` category is split into six categories (`c10d`, `rpc`, `ddp`, `pipeline`, `fsdp`, `sharded`)

And optionally, assign *one* of the following topic labels:

- `topic: bc_breaking`
- `topic: deprecation`
- `topic: new feature`
- `topic: improvements`
- `topic: bug fixes`
- `topic: performance`
- `topic: documentation`
- `topic: developer feature`
- `topic: non-user visible` <- assign this label if you think PR should not be mentioned in release notes

Later assigning such labels might become a mandatory pre-merge check

cc @brianjo @mruberry @seemethere @malfet @pytorch/pytorch-dev-infra

**Comment by malfet** [comment]
@anjali411 please feel free to edit the issue and add short description to every topic and category

**Comment by anjali411** [comment]
We should also add a check that each PR is assigned with a **unique** category and topic label. This is important, because each PR will ultimately go under the assigned category/topic in the release notes, and we should do this work upfront because the commits will be distributed to the PoCs based on the category label.

**Comment by NicolasHug** [comment]
Courtesy of @pmeier  we have a bot in torchvision that pings the merger of a PR if they forgot to add an appropriate label: https://github.com/pytorch/vision/blob/main/.github/workflows/pr-labels.yml and https://github.com/pytorch/vision/blob/main/.github/process_commit.py

It might need to be adapted for pytorch core because of the different merging process (at least until core switches to github-first).

You can see it in action e.g. here https://github.com/pytorch/vision/pull/4618#issuecomment-943586489

**Comment by malfet** [comment]
I'm thinking about refactoring this logic into probot and then enable it for all the projects.

**Comment by suo** [comment]
For each group we have oncall labels, module labels, and now release notes labels. Could we consolidate them into one? What's the case for release notes labels to be separate?

**Comment by malfet** [comment]
@suo, for the ones that you know label already exists, can you please edit issue description and add alias (and we definitely should fold)

**Comment by suo** [comment]
I added a few that I know of or could find through basic searching. One thing I noticed is that some of the older labels are `oncall:`, which predates the `module:` system. I would be in favor of consolidating on the `module:` system and using that not only for release notes but future CODEOWNERS stuff, but it's not relevant for now.

**Comment by suo** [comment]
Also, what do you think about doing something like: https://www.conventionalcommits.org/. The advantage being that it's a common thing and tooling exists for all the various components (linting, release note generation, etc.). The disadvantage being that it's commit based (although personally I would prefer just naming my commit instead of clicking around GH).

**Comment by anjali411** [comment]
> 

I think that's a fair critique. I'd say that we should probably combine the module and release note labels, however we should ensure that people only add one module label on a PR. The reason is that if we add multiple module labels on the PR, then it wouldn't be useful for release notes categorization.

**Comment by gchanan** [comment]
> I added a few that I know of or could find through basic searching. One thing I noticed is that some of the older labels are oncall:, which predates the module: system. I would be in favor of consolidating on the module: system and using that not only for release notes but future CODEOWNERS stuff, but it's not relevant for now.

I don't believe `oncall:` predates the `module:` system, or if they do it doesn't matter.  The point of the `oncall:` labels are to help the core triagers -- it's an indication that another oncall besides the core one should handle it and the core oncall doesn't need to worry about further labeling, marking triage review, etc.

**Comment by malfet** [comment]
I'm to modify probot to assign label watchers  as reviewers on PR

**Comment by anjali411** [comment]
Triage discussion notes:

1. Add separate `release_notes: ...` labels instead of using the existing module labels. This would allow us to enforce having a unique label on each PR, while allowing people to add multiple module labels (which separately should add relevant people as reviewers and not cc them)
2. Make it mandatory for people to add a *unique* release note category label *as well as a unique* topic to every PR except when they add a `non-user facing` topic, i.e., prevent people from merging a PR unless the above stated condition is met. 
3. Probot should link people to this issue or a new issue containing a list of release_note category and topic labels. People are welcome to add new release note categories if they like. While adding a new category, keep in mind how you'd like to publish the change in the final release notes. Is it a big enough feature to be its own category? Can the change in question fit under an existing category in the final release notes?

**Comment by anjali411** [comment]
should we keep this issue open?

**Comment by Gamrix** [comment]
Is there any way to add this to the git commit template for Pytorch so that we can add it when making a commit?

**Comment by ZolotukhinM** [comment]
Would it be possible to somehow make these labels optional or automatic for some parts of the code? For instance, when working on JIT/NNC most of the changes are internal and thus do not need a release notes entry. Adding friction to already cumbersome developer workflow seems not ideal here, it will discourage people from splitting changes to smaller commits for instance.

I agree that writing release notes is not a one-person task, but I'm also trying to not regress other aspects of developer efficiency. What other options have been considered? What if before the release we send out email to all people with commits in the given time period asking them to prepare release notes for **their** commits?

**Comment by suo** [comment]
I do wonder if this change needs more socialization before going forward, as it proposes to change the workflow for every PyTorch developer. The current implementation of the soft mandate also leaves something to be desired; it adds noise and friction to the workflow without giving users options that could reduce that friction.

I could imagine a couple different mitigations:
- Invest more in @anjali411's script to perform post-hoc labeling. The PR indicates coverage was 63% for PyTorch 1.10, which seems pretty good. Are we convinced that further improvements are not possible? If we could get the coverage high enough, we could potentially not change the workflow at all.
- Perhaps use the script as a first pass, so most labels could be applied automatically. This would at least reduce noise/friction in most cases.
- Or add to the labeling workflow some way for people to configure what labels will be applied automatically, similar to the ccbot tracking issue.
- In GHF world, organize PRs by module (by codeowners) and require that module maintainers write release notes.

**Comment by anjali411** [comment]
- Perhaps use the script as a first pass, so most labels could be applied automatically. This would at least reduce noise/friction in most cases.

This is a pretty cool idea. We should probably consider doing a conditional ping based on whether the script was able to auto categorize the commit. However, currently there's no logic in the script to add topics (besides bc-breaking and deprecation which are just inferred from the corresponding labels). We thought it might be easier to topic the PR while merging it since they'd remember the nits and it'd be anyway more accurate than one person looking through hundreds of commits and deciding which topic is suitable for each PR (common case for JIT, Quantization, Release Engineering teams).

My one concern for the script is that it can easily become outdated (if the existing files/folders are added or new files/folders are removed). would it make sense to ping people to consider updating the script when they are adding or removing files? 

cc. @bdhirsh who is working on the release notes this time.

**Comment by anjali411** [comment]
@ZolotukhinM I understand your concerns and agree that the current workflow might be a bit noisy or disruptive, especially for people who are working on aspects of features that are mostly non user facing. As a first step alleviate this, I think we should take @suo 's suggestion to try to automatically label the PRs. As for JIT/NNC, I have noticed that there's usually an `oncall:jit` label. Is it automatically triggered? should we just use that to tag the PR? What do you think would be a good strategy to topic the PRs? Ansley and some other people who were responsible for compiling JIT release notes in the previous releases shared that it was a lot of work for one person to topic hundreds of PRs, especially because it was difficult to figure out what the PR was doing (since there was barely enough PR description among other things). So, I think we should consider adding topics upfront in some way. We could add some logic to automatically topic some PRs as non user facing based on files changed or labels if you have suggestions. You are also welcome to create a PR updating the script directly!

**Comment by atalman** [comment]
@suo and @anjali411  may I suggest we start a new task and create small functional spec on how these new modifications will look like. Once we have these specs we will post it  on the workplace and make sure we get by in from all the parties before continuing. I really like @suo suggestions I think we should definitely do something more automatic in terms of labelling.

**Comment by garymm** [comment]
It seems some bot is leaving comments asking for labels to be added even when they are present.

Example: <https://github.com/pytorch/pytorch/pull/72883#issuecomment-1054564487>

The PR already has the requested labels:
![image](https://user-images.githubusercontent.com/421339/156061238-486b5b82-1d59-453a-b197-541ac491ec47.png)

Please fix!

**Comment by NicolasHug** [comment]
@anjali411 @malfet is it OK for me to add a `release notes: hub` label for torchhub-related PRs? CC @vmoens

**Comment by anjali411** [comment]
@NicolasHug yeah absolutely. you would just need to add a new github label and you will be good to go

**Comment by malfet** [comment]
This has been finished, closing
